core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java

```diff
@@ -61,16 +69,29 @@ public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceO
     public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceInput)
     {
         int positionCount = sliceInput.readInt();
+        int notNullPositionCount = sliceInput.readInt();
 
         int[] offsets = new int[positionCount + 1];
-        sliceInput.readBytes(Slices.wrappedIntArray(offsets), SIZE_OF_INT, positionCount * SIZE_OF_INT);
+        int[] lengths = new int[notNullPositionCount];
+        sliceInput.readBytes(Slices.wrappedIntArray(lengths), 0, notNullPositionCount * SIZE_OF_INT);
 
         boolean[] valueIsNull = decodeNullBits(sliceInput, positionCount).orElse(null);
 
         int blockSize = sliceInput.readInt();
         Slice slice = Slices.allocate(blockSize);
         sliceInput.readBytes(slice);
 
+        int notNullPosition = 0;
+        int offset = 0;
+
+        offsets[0] = 0;
+        for (int i = 1; i < positionCount + 1; i++) {
+            if (valueIsNull == null || !valueIsNull[i - 1]) {
```

Comment:
nit: we could potentially try a branchless version e.g. something like:
```
            boolean isNull = valueIsNull == null | !valueIsNull[i - 1];
            offset += isNull ? 0 : lengths[notNullPosition];
            notNullPosition += isNull ? 0 : 1;
            offsets[i] = offset;
```

Reply:
good point